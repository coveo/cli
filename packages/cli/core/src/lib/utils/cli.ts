import {confirm} from '@coveo/cli-commons/utils/ux';

import globalConfig from '@coveo/cli-commons/config/globalConfig';
import {buildEvent} from '@coveo/cli-commons/analytics/eventUtils';

/**
 *
 * @param message confirmation prompt (yes/no)
 * @param questionName Used for analytics purposes to name the action. Should not contain sensitive information.
 * See https://coveord.atlassian.net/wiki/spaces/RD/pages/2214168207/Amplitude for taxonomy
 * @returns
 */
export async function confirmWithAnalytics(
  message: string,
  questionName: string
): Promise<boolean> {
  const doAction = await confirm(message, false);
  if (doAction) {
    trackEvent(`confirmed ${questionName}`);
    return true;
  } else {
    trackEvent(`cancelled ${questionName}`);
    return false;
  }
}

async function trackEvent(questionName: string) {
  await globalConfig.get().runHook('analytics', {
    event: buildEvent(questionName, {}),
  });
}
