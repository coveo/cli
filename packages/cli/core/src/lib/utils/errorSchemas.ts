import {z} from 'zod';

export const isAggregatedErrorLike = (error: any): error is AggregateError =>
  AggregateErrorLike.safeParse(error).success;

const AggregateErrorLike = z.object({
  name: z.literal('AggregateError'),
  errors: z.array(z.string()),
  message: z.string(),
});

export const isErrorLike = (error: any): error is Error =>
  ErrorLike.safeParse(error).success;

const ErrorLike = z.object({
  name: z.string(),
  message: z.string(),
});
