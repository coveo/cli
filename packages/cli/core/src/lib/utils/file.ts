import {readdirSync} from 'fs';
import {extname, join} from 'path';

export function fileDepthSearch(
  dirPath: string,
  fileExtension = '.json'
): string[] {
  const filePaths: string[] = [];
  const files = readdirSync(dirPath, {withFileTypes: true});
  for (const file of files) {
    const filePath = join(dirPath, file.name);
    if (file.isDirectory()) {
      filePaths.push(...fileDepthSearch(filePath));
    }
    if (file.isFile() && extname(filePath) === fileExtension) {
      filePaths.push(filePath);
    }
  }

  return filePaths;
}
