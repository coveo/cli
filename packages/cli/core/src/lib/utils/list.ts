export function without<T>(array: T[], values: T[]): T[] {
  return array.filter((field) => !values.includes(field));
}
