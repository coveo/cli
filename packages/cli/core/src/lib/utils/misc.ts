import {resolve} from 'path';
import {valid} from 'semver';

export function getPackageVersion(packageName: string) {
  const pathToPackageJson = resolve(
    __dirname,
    ...new Array(3).fill('..'),
    'package.json'
  );
  const pkg = require(pathToPackageJson);
  const dep =
    pkg.dependencies[packageName] ||
    pkg.devDependencies[packageName] ||
    pkg.peerDependencies[packageName];
  const defaultVersion = valid(dep?.toString())!;

  return defaultVersion;
}
