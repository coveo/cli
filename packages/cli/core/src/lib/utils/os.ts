export const appendCmdIfWindows = (cmd: TemplateStringsArray) =>
  `${cmd}${process.platform === 'win32' ? '.cmd' : ''}`;
