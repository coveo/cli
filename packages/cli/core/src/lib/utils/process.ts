import {spawn, SpawnOptions, ChildProcess} from 'child_process';

function isErrnoException(error: Error): error is NodeJS.ErrnoException {
  return Object.hasOwnProperty.call(error, 'errno');
}

/**
 *
 * @private
 * @param {string} command The command to run
 * @param {string[]} args List of string arguments
 * @param {SpawnOptions} [options={}]
 * @returns {Promise<number>} A promise of the the child process exit code
 */
export function spawnProcess(
  command: string,
  args: string[],
  options: SpawnOptions = {}
): Promise<number> {
  return new Promise((resolve, reject) => {
    spawn(command, args, {
      stdio: ['inherit', 'inherit', 'inherit'],
      ...options,
    }).on('close', (code) => {
      if (code !== 0) {
        // TODO: inject logger decorator so this util method can log
        // this.debug(
        //   `Following command has failed: ${command} ${args.join(' ')}`
        // );

        // Any error message triggered by the execution of the subprocess will be
        // displayed since the terminal is shared by both parent and child processes.
        // So no need to bubble up the error.
        reject(code);
      } else {
        resolve(0);
      }
    });
  });
}

export interface SpawnProcessOutput {
  stdout: string;
  stderr: string;
  exitCode?: string;
}

export async function spawnProcessOutput(
  command: string,
  args: string[],
  options: SpawnOptions = {}
): Promise<SpawnProcessOutput> {
  return new Promise((resolve) => {
    const output: SpawnProcessOutput = {
      stdout: '',
      stderr: '',
    };

    const child = spawn(command, args, options);

    child.stdout?.on('data', (d) => {
      output.stdout += d;
    });

    child.stderr?.on('data', (d) => {
      output.stderr += d;
    });

    const closeListener = (code: number | null) => {
      output.exitCode = Number.isInteger(code) ? String(code) : undefined;
      resolve(output);
    };

    child.on('close', closeListener);

    child.on('error', (e) => {
      output.stderr += e.message;
      if (isErrnoException(e) && e.code) {
        child.off('close', closeListener);
        output.exitCode = e.code;
        resolve(output);
      }
    });
  });
}

export const handleForkedProcess = (subprocess: ChildProcess) => {
  return new Promise<void>((resolve, reject) => {
    subprocess.on('exit', () => {
      resolve();
    });
    subprocess.on('message', (message) => {
      reject(message);
    });
  });
};
