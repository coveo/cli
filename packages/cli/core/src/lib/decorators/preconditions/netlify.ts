import {appendCmdIfWindows} from '../../utils/os';
import {getBinVersionPrecondition} from './binPreconditionsFactory';

const angularCliVersionMatcher = /^netlify-cli\/(?<version>\d+\.\d+\.\d+).*$/m;

export const IsNetlifyCliVersionInRange = getBinVersionPrecondition(
  appendCmdIfWindows`netlify`,
  {
    prettyName: 'Netlify-CLI',
    howToInstallBinText:
      'You can install the Netlify-CLI by running npm i -g netlify-cli',
    installLink: 'https://docs.netlify.com/cli/get-started/#installation',
    params: ['version'],
  },
  (stdout: string) =>
    stdout.match(angularCliVersionMatcher)?.groups?.['version'] ?? ''
);
