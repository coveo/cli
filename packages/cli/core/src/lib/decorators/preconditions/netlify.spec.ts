jest.mock('../../utils/process');
jest.mock('../../utils/misc');
jest.mock('../../utils/os');

import {spawnProcessOutput} from '../../utils/process';
import {getFakeCommand} from '@coveo/cli-commons/utils/getFakeCommand';

import {IsNetlifyCliVersionInRange} from './netlify';
import {CliUx} from '@oclif/core';
import {appendCmdIfWindows} from '../../utils/os';
import {getPackageVersion} from '../../utils/misc';
import {fancyIt} from '@coveo/cli-commons-dev/testUtils/it';
import {CLICommand} from '@coveo/cli-commons/command/cliCommand';

describe('IsNgInstalled', () => {
  const mockedSpawnProcessOutput = jest.mocked(spawnProcessOutput);
  const mockedGetPackageVersion = jest.mocked(getPackageVersion);
  const mockedAppendCmdIfWindows = jest.mocked(appendCmdIfWindows);

  const mockAppendCmdIfWindows = () => {
    mockedAppendCmdIfWindows.mockImplementationOnce((input) => `${input}`);
  };

  const mockConfirm = () => {
    Object.defineProperty(CliUx.ux, 'confirm', {value: jest.fn()});
  };
  let fakeCommand: CLICommand;

  beforeAll(() => {
    mockConfirm();
    mockAppendCmdIfWindows();
  });

  beforeEach(() => {
    jest.resetAllMocks();
    fakeCommand = getFakeCommand();
    mockedGetPackageVersion.mockReturnValue('1.0.0');
  });

  describe('when the requiredVersion is not a semver valid string', () => {
    fancyIt()('should throw', async () => {
      const fakeCommand = getFakeCommand();

      await expect(
        IsNetlifyCliVersionInRange('foo')(fakeCommand)
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('when netlify-cli is not installed', () => {
    beforeEach(() => {
      mockedSpawnProcessOutput.mockResolvedValue({
        exitCode: 'ENOENT',
        stderr: '',
        stdout: '',
      });
    });

    fancyIt()('should throw', async () => {
      const fakeCommand = getFakeCommand();
      await expect(
        IsNetlifyCliVersionInRange('>=0.0.1')(fakeCommand)
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('when an unknown error happens while checking for netlify-cli', () => {
    beforeEach(() => {
      mockedSpawnProcessOutput.mockResolvedValue({
        exitCode: '1',
        stderr: 'some random error oh no',
        stdout: '',
      });
    });

    fancyIt()('should return false and warn', async () => {
      await expect(
        IsNetlifyCliVersionInRange('>=0.0.1')(fakeCommand)
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('when the installed version of netlify-cli is lower than the required one', () => {
    beforeEach(() => {
      mockedSpawnProcessOutput.mockResolvedValue({
        exitCode: '0',
        stderr: '',
        stdout: 'netlify-cli/0.9.0',
      });
    });

    fancyIt()('should throw', async () => {
      const fakeCommand = getFakeCommand();

      await expect(
        IsNetlifyCliVersionInRange('>=1.0.0')(fakeCommand)
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('when the installed version of netlify-cli is above than the required one', () => {
    beforeEach(() => {
      mockedSpawnProcessOutput.mockResolvedValue({
        exitCode: '0',
        stderr: '',
        stdout: 'netlify-cli/1.1.0',
      });
    });

    fancyIt()('should not throw', async () => {
      const fakeCommand = getFakeCommand();

      await expect(
        IsNetlifyCliVersionInRange('>=1.0.0')(fakeCommand)
      ).resolves.not.toThrow();
    });
  });
});
