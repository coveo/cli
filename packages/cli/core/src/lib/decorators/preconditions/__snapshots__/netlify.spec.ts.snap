// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsNgInstalled when an unknown error happens while checking for netlify-cli should return false and warn 1`] = `
"foo requires a valid Netlify-CLI installation to run.
An unknown error happened while running undefined version.
some random error oh no

You can install the Netlify-CLI by running npm i -g netlify-cli

Please visit https://docs.netlify.com/cli/get-started/#installation for more detailed installation information."
`;

exports[`IsNgInstalled when netlify-cli is not installed should throw 1`] = `
"foo requires Netlify-CLI to run.

You can install the Netlify-CLI by running npm i -g netlify-cli

Please visit https://docs.netlify.com/cli/get-started/#installation for more detailed installation information."
`;

exports[`IsNgInstalled when the installed version of netlify-cli is lower than the required one should throw 1`] = `
"foo needs a Netlify-CLI version in this range: ">=1.0.0"
Version detected: 0.9.0

You can install the Netlify-CLI by running npm i -g netlify-cli

Please visit https://docs.netlify.com/cli/get-started/#installation for more detailed installation information."
`;

exports[`IsNgInstalled when the requiredVersion is not a semver valid string should throw 1`] = `
"Required version invalid: "foo".
Please report this error to Coveo: https://github.com/coveo/cli/issues/new"
`;
